cranelift/codegen/src/machinst/mod.rs

```diff
@@ -100,20 +81,18 @@ pub use inst_common::*;
 pub mod valueregs;
 pub use reg::*;
 pub use valueregs::*;
-pub mod debug;
-pub use regmapping::*;
-pub mod regmapping;
 pub mod reg;
 
 /// A machine instruction.
 pub trait MachInst: Clone + Debug {
     /// Return the registers referenced by this machine instruction along with
     /// the modes of reference (use, def, modify).
-    fn get_regs(&self, collector: &mut RegUsageCollector);
+    fn get_operands<F: Fn(VReg) -> VReg>(&self, collector: &mut OperandCollector<'_, F>);
 
-    /// Map virtual registers to physical registers using the given virt->phys
-    /// maps corresponding to the program points prior to, and after, this instruction.
-    fn map_regs<RUM: RegUsageMapper>(&mut self, maps: &RUM);
+    /// Get the clobbers for this instruction, if any. These are like
+    /// defs (written registers) but are invalid, i.e., cannot legally
+    /// be read.
+    fn get_clobbers(&self) -> &[Writable<Reg>];
```

Comment:
I was starting to write an answer here discussing allocations and how I'd prefer to avoid them, and the tradeoffs vs flexibility and maybe `Cow` but that's awkward, etc., then realized we've been down this road before -- we have all the same tradeoffs for ordinary operands. So I think maybe the better answer is to actually go through the operand-collector interface (`reg_clobber` alongside `reg_use`, `reg_def` and friends), what do you think?

Reply:
Done! This is also actually dead right now, as calls use phantom defs (a legacy from regalloc.rs). I'll create a followup issue to track using actual clobbers, which should be very marginally faster.

Reply:
#4027